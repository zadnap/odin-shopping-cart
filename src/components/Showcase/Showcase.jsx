import styles from './Showcase.module.scss';
import { useEffect, useState } from 'react';
import { Link } from 'react-router-dom';
import Loader from '@/components/Loader/Loader';
import ErrorMessage from '@/components/ErrorMessage/ErrorMessage';

const API_KEY = import.meta.env.VITE_TMDB_API_KEY;

function Showcase() {
  const [movie, setMovie] = useState(null);
  const [errorMessage, setErrorMessage] = useState(null);

  useEffect(() => {
    const fetchMovie = async () => {
      try {
        const [movieRes, genresRes] = await Promise.all([
          fetch(
            `https://api.themoviedb.org/3/movie/popular?api_key=${API_KEY}&language=en-US&page=1`
          ),
          fetch(
            `https://api.themoviedb.org/3/genre/movie/list?api_key=${API_KEY}&language=en-US`
          ),
        ]);

        if (!movieRes.ok || !genresRes.ok)
          throw new Error(
            (movieRes.status === 404
              ? 'Could not find popular movies. '
              : 'Something went wrong while fetching movie data. ') +
              'Please try again later.'
          );

        const movieData = await movieRes.json();
        const genresData = await genresRes.json();

        const result = movieData.results[0];
        const genreList = genresData.genres;

        setMovie({
          id: result.id,
          title: result.title,
          desc: result.overview,
          tags: [
            result.release_date.slice(0, 4),
            ...result.genre_ids
              .map((id) => genreList.find((g) => g.id === id)?.name)
              .filter(Boolean),
          ],
          backdropSrc: `https://image.tmdb.org/t/p/original${result.backdrop_path}`,
        });
      } catch (error) {
        setErrorMessage(error.message);
      }
    };

    fetchMovie();
  }, []);

  return (
    <div className={styles.showcaseContainer}>
      {errorMessage ? (
        <ErrorMessage message={errorMessage} />
      ) : movie ? (
        <article
          className={styles.showcase}
          style={{
            backgroundImage: `url(${movie.backdropSrc})`,
          }}
        >
          <ul className={styles.tagList}>
            {movie.tags.map((tag) => (
              <li key={tag} className={styles.tag}>
                {tag}
              </li>
            ))}
          </ul>
          <div className={styles.movieInfo}>
            <h1 className={styles.title}>{movie.title}</h1>
            <p className={styles.desc}>
              <span className={styles.clampText}>{movie.desc}</span>{' '}
              <Link className={styles.seeMore} to={`/movies/${movie.id}`}>
                See more
              </Link>
            </p>
          </div>
        </article>
      ) : (
        <Loader />
      )}
    </div>
  );
}

export default Showcase;
