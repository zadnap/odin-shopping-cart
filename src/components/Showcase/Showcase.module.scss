@use '@/styles/variables.scss' as *;
@use '@/styles/keyframes.scss' as *;

.showcaseContainer {
  height: 400px;
}

.showcase {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  padding: $gap;
  border-radius: $border-radius;
  background-size: cover;
  background-repeat: no-repeat;
  height: 100%;
  animation: fadeIn 0.2s ease;

  .tagList {
    display: flex;
    gap: 7px;

    .tag {
      font-size: 1.4rem;
      border-radius: 50px;
      background-color: $overlay-color;
      padding: 7px 15px;
    }
  }

  .movieInfo {
    display: flex;
    flex-direction: column;
    gap: 5px;
    background-color: $overlay-color;
    padding: 7px 12px;
    border-radius: $border-radius;

    .title {
      font-size: 2rem;
      font-weight: 700;
      text-shadow: 1px 1px 2px rgba(0, 0, 0, 0.5);
    }
    .desc {
      display: flex;

      .clampText {
        white-space: nowrap;
        overflow: hidden;
        text-overflow: ellipsis;
        max-width: calc(100% - 80px);
      }
      .seeMore {
        color: $subtext-color;
        transition: color 0.1s;

        &:hover {
          color: $text-color;
          font-weight: 500;
          text-decoration: underline;
        }
      }
    }
  }
}
