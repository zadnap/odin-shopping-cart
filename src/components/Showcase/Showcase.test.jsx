import { describe, it, expect } from 'vitest';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import Showcase from './Showcase';

const mockProps = {
  id: 1,
  title: 'How to Train Your Dragon',
  backdropSrc:
    'https://image.tmdb.org/t/p/original/7HqLLVjdjhXS0Qoz1SgZofhkIpE.jpg',
  tags: ['Action', 'Movie', '2025', '6+'],
  desc: 'On the rugged isle of Berk, where Vikings',
};

describe('Showcase component', () => {
  it('should render to match snapshot', () => {
    const { container } = render(
      <MemoryRouter>
        <Showcase {...mockProps} />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });
});
