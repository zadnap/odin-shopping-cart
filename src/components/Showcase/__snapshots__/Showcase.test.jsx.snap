// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Showcase component > should render to match snapshot 1`] = `
<div>
  <div
    class="_showcaseContainer_43ad9d"
  >
    <div
      aria-busy="true"
      aria-live="polite"
      class="_loader_0c7d59"
      role="status"
    >
      <div
        class="_loadingCircle_0c7d59"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-video _icon_0c7d59"
          data-icon="video"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 576 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 128C0 92.7 28.7 64 64 64l256 0c35.3 0 64 28.7 64 64l0 256c0 35.3-28.7 64-64 64L64 448c-35.3 0-64-28.7-64-64L0 128zM559.1 99.8c10.4 5.6 16.9 16.4 16.9 28.2l0 256c0 11.8-6.5 22.6-16.9 28.2s-23 5-32.9-1.6l-96-64L416 337.1l0-17.1 0-128 0-17.1 14.2-9.5 96-64c9.8-6.5 22.4-7.2 32.9-1.6z"
            fill="currentColor"
          />
        </svg>
      </div>
      <p
        class="_message_0c7d59"
      >
        Just a moment
      </p>
    </div>
  </div>
</div>
`;
